File: inst/dynload/C/mymod.c
/* compile with R CMD SHLIB mymod.c or (on Windows) Rcmd SHLIB mymod.c */
/* Example from lsoda documentation */
#include <R.h> /* gives F77_CALL through R_ext/RS.h */

static double parms[3];

/* A trick to keep up with the parameters */
#define k1 parms[0]
#define k2 parms[1]
#define k3 parms[2]

/* initializer */
void mymod(void (* odeparms)(int *, double *))
{
    int N=3;
    odeparms(&N, parms);
}

/* Derivatives */
void myderivs(int *neq, double *t, double *y, double *ydot)
{
    ydot[0] = -k1*y[0] + k2*y[1]*y[2];
    ydot[2] = k3 * y[1]*y[1];
    ydot[1] = -ydot[0]-ydot[2];
}

/* Jacobian */
void myjac(int *neq, double *t, double *y, int *ml,
                  int *mu, double *pd, int *nrowpd)
{
    pd[0] = -k1;
    pd[1] = k1;
    pd[2] = 0.0;
    pd[(*nrowpd)] = k2*y[2];
    pd[(*nrowpd) + 1] = -k2*y[2] - 2*k3*y[1];
    pd[(*nrowpd) + 2] = 2*k3*y[1];
    pd[(*nrowpd)*2] = k2*y[1];
    pd[2*(*nrowpd) + 1] = -k2 * y[1];
    pd[2*(*nrowpd) + 2] = 0.0;
}
/* End of example */
 

File: inst/dynload/Fortran/mymod.f
c -------- mymod.f -> mymod.so (or dll or whatever) ------
c compile with R CMD SHLIB mymod.f or (on Windows) Rcmd SHLIB mymod.f

c copied directly from the lsoda documentation,with common block added

      subroutine mymod(odeparms)
      external odeparms
      integer N
      double precision parms(3)
      common /myparms/parms

      N = 3
      call odeparms(N, parms)
      return
      end

      subroutine myderivs (neq, t, y, ydot)
      double precision t, y, ydot, parms(3)
      integer neq
      dimension y(3), ydot(3)
      common /myparms/parms

      ydot(1) = -parms(1)*y(1) + parms(2)*y(2)*y(3)
      ydot(3) = parms(3)*y(2)*y(2)
      ydot(2) = -ydot(1) - ydot(3)
      return
      end

      subroutine myjac (neq, t, y, ml, mu, pd, nrowpd)
      integer neq, ml, mu, nrowpd
      double precision y(*), pd(nrowpd,*), t, parms(3)
      common /myparms/parms

      pd(1,1) = -parms(1)
      pd(2,1) =  parms(1)
      pd(3,1) = 0.0
      pd(1,2) = parms(2)*y(3)
      pd(2,2) = -parms(2)*y(3) - 2*parms(3)*y(2)
      pd(3,2) = 2*parms(3)*y(2)
      pd(1,3) = parms(2)*y(2)
      pd(2,3) = -parms(2)*y(2)
      pd(3,3) = 0.0

      return
      end

 
